File: apps/android/app/src/main/java/com/healthmd/domain/practice/PracticeFeaturePolicy.kt
package com.healthmd.domain.practice

import com.healthmd.BuildConfig

/**
 * Fail-closed rollout boundary for additive Health.md Practice functionality.
 *
 * Packaging future Practice code is necessary but never sufficient. A separately governed
 * qualification attestation must match [QUALIFICATION_VERSION]. No production attestation is
 * supplied today, so the existing local-first, account-free application remains unchanged.
 */
data class PracticeFeaturePolicy(
    val compiledIn: Boolean,
    val qualificationAttested: Boolean,
) {
    val isEnabled: Boolean = compiledIn && qualificationAttested

    companion object {
        const val QUALIFICATION_VERSION = "practice-v1-qualified"

        fun current(): PracticeFeaturePolicy = PracticeFeaturePolicy(
            compiledIn = BuildConfig.PRACTICE_COMPILED_IN,
            qualificationAttested = qualificationIsValid(null),
        )

        fun resolve(
            compiledInValue: String?,
            qualificationValue: String?,
        ): PracticeFeaturePolicy = PracticeFeaturePolicy(
            compiledIn = compiledInValue == "included",
            qualificationAttested = qualificationIsValid(qualificationValue),
        )

        private fun qualificationIsValid(value: String?): Boolean =
            value == QUALIFICATION_VERSION
    }
}

File: apps/apple/HealthMd/Shared/Practice/PracticeFeaturePolicy.swift
import Foundation

/// Fail-closed rollout boundary for additive Health.md Practice functionality.
///
/// Compiling future Practice code into an app is necessary but never sufficient to expose it.
/// A separately governed qualification attestation must also match the reviewed gate version.
/// The production app intentionally supplies no attestation today, so existing local-first,
/// account-free behavior is unchanged.
nonisolated struct PracticeFeaturePolicy: Equatable, Sendable {
    static let qualificationVersion = "practice-v1-qualified"

    let compiledIn: Bool
    let qualificationAttested: Bool

    var isEnabled: Bool { compiledIn && qualificationAttested }

    static var current: PracticeFeaturePolicy {
        PracticeFeaturePolicy(
            compiledIn: compileTimeIncluded,
            qualificationAttested: qualificationIsValid(nil)
        )
    }

    static func resolve(
        compiledInValue: String?,
        qualificationValue: String?
    ) -> PracticeFeaturePolicy {
        PracticeFeaturePolicy(
            compiledIn: compiledInValue == "included",
            qualificationAttested: qualificationIsValid(qualificationValue)
        )
    }

    private static func qualificationIsValid(_ value: String?) -> Bool {
        value == qualificationVersion
    }

    private static var compileTimeIncluded: Bool {
#if HEALTHMD_PRACTICE_COMPILED_IN
        true
#else
        false
#endif
    }
}
